settling_messages_when_receiver_is_closing.md

```diff
@@ -0,0 +1,104 @@
+to close the Receiver.
+
+In all cases this proposal only applies to Receivers that were opened in
+`peekLock` mode.
+
+## How our current design works
+
+Service Bus has some situations where settling on the receiver
+where the original message was received is important.
+
+Most of the time this works out transparently to the user - a typical
+usage pattern for Service Bus would be:
+
+```typescript
+receiver.subscribe({
+  processMessage: async (msg) => {
+    // do something with `msg`
+  },
+  // ... other parts elided for example.
+  {
+    // this is actually the default
+    autoComplete: true
+  }
+});
+```
+
+In this case if we do not throw an Error in `processMessage` the
+message will be automatically completed for the user.
+
+## Where our current design falls down
+
+However, some customers do something more akin to this:
+
+```typescript
+
+const someOutsideStoreOfMessages = [];
+
+receiver.subscribe({
+  processMessage: async (msg) => {
+    someOutsideStoreOfMessages.push(msg);
+  },
+  // ... other parts elided for example.
+  {
+    // ie: I want to handle completing the message
+    autoComplete: false
+  }
+})
+
+// ...in another part of their application
+
+// by some other criteria decide to commit the messages
+for (const msg of someOutsideStoreOfMessages) {
+  await msg.complete();
+}
+
+```
+
+In this model users must now make sure that the `receiver` remains open
+until they complete their last message.
+
+If the user closes the Receiver to "stop" the subscription they will now run into the
+problem where they must have the original Receiver to complete any outstanding
+messages.
+
+If they leave it open then they will continue to receive any other messages, making it
+difficult to properly stop.
+
+## Design ideas
+
+Some options to consider:
+
+1. Make `close()` more intelligent.
+
+   Currently `close()` just shuts things down without any regard for messages that are outstanding (ie,
+   reasons to keep the link open). One option is to add a configuration parameter that would allow
+   the user to say "close but ignore all connections" or "close _after_ any outstanding messages are settled".
+
+   The default would be "close but ignore all connections" in order to avoid any surprises for users that expect
+   close() to be immediate. Advanced users can change this behavior.
+
+   ```typescript
+   receiver.close({
+     afterMessagesSettled: true,
+   });
+   ```
+
+2. Add a `stop()` method
+
+   We could make it possible for users to stop (ie, `pause`) a receiver (meaning - remove credits and subscription
+   event handlers).
+
+   However, if the 'stop()' is generally followed by a `close()` then we're probably better off with option #1.
+
+   As part of different PR for track 1 we are adding in an `open()` call which might pair nicely with this.
+
+3. Return a subscription instance from `subscribe` that provides methods for the user to "stop" the subscription.
```

Comment:
I vote for Option 3 i.e return a subscription from `subscribe` that has methods to stop receiving any more messages.

Option 2 is a non-starter for me as 
- there would be confusion as to why one would need a `stop()` when there is a `close()`
- The `stop()` might be expected to be applied for other operations on the receiver

Option 1: This tells me that I can have the receiver closed after messages are settled. It doesn't tell me that no more messages can be received.

Option 3 makes it very clear as to what the stop() would do and what its scope of influence is. Even without the problem of settling messages but no more new messages, I would advocate for Option 3 for all the same reasons as to why it existed in EH.

Reply:
I'm not a big fan of #2 either, primarily because the only use cases we've heard for a "pause" is actually them really wanting to close the receiver.

As to your other points:
> Option 3 makes it very clear as to what the stop() would do and what its scope of influence is. Even without the problem of settling messages but no more new messages, I would advocate for Option 3 for all the same reasons as to why it existed in EH.

I don't disagree with this. 

However, if the user is just trying to shut down the Receiver overall then providing the subscription.stop() is just more work. It seems like option #1 gives them a very clear way of doing both the "abort now!" vs "shutdown cleanly" without regard to the operation they're using (ie, subscribe, receiveBatch, etc..) which is a simpler model.

Reply:
> However, if the user is just trying to shut down the Receiver overall then providing the subscription.stop() is just more work.

The user can always call `receiver.close()` in such case. They don't need to call `stop()` on the subscription first and then `close()` on the receiver.

Reply:
Oh, of course - all the solutions are equivalent from a feature standpoint. :)

This was more about saving the user a step if they're basically always doing the two together (ie, really they're just shutting down, not stopping and then hoping to restart).

Talking with @bterlson there are some implications (as we thought out loud) that making 'Receiver.close()' be more intelligent might also imply that all the other close()'s also get more intelligent. So we might be pushing user's expectations towards "well, what if I have outstanding Receivers or Senders, etc...". That's not a road we want to go down.

So I think that option #2 is our best and simplest shot. This particular feature is probably not the right time to break new ground :)

Reply:
Also, just to note - given that we're following what EventHubs is doing the method should be called `close()`, not `stop()`, on the returned Subscription object.